### ISSUE TITLE
вопросы к заданию для самостоятельной работе (Лекция 9)

### ISSUE THREAD

**Issue by Poodelay** [kalininei/CFDCourse25#3]
Здравствуйте, Евгений Игоревич, у меня несколько вопросов по домашней работе
Учитывает ли метод linear_combination(iface) в коэффициентах, которые возвращает, что производная дальше умножается на площадь грани, или перед добавлением этих коэффициентов в матрицу их нужно домножать на площадь грани?
Возможно ли, что выше указанный метод может вернуть INVALID_INDEX? я условие, что индексы i и j являются центрами ячеек, проверяю, сравнивая их с icell_positive и icell_negative, которые как раз могут принимать значение INVALID_INDEX
Правильно ли я понимаю, что в методе approximate_lhs должно быть только два цикла, описанные в pdf файле под номерами (9.5) и (8.2)
Извините за такое количество вопросов, у меня как-то совсем не выходит с этим заданием

**Comment by kalininei** [comment]
Отвечая на ваши вопросы:
- нет, linear_combination не умножает производную на площадь грани
- нет, linear_combination не может вернуть невалидный индекс. Он возвращает пары: индекс коллокации и его вес.
- проверять, что индекс коллокации `i` является центром ячейки можно по его значению: по построению индексов у нас сначала идут центры ячеек, потом центры граничных граней. Поэтому, если `i<Ncells`, то это центр ячейки 
- да, поскольку г.у. первого рода, то собираться левая часть должна в два цикла: (9.5) для уравнений для ячеек и (8.2) для границ

А вообще, ничто не мешает вам посмотреть реализацию любого метода, в котором вы сомневаетесь.
Построение нормальной производной мы и на паре обсудили и в бумажке она описана. Поэтому реализация linear_combination должна быть вам абсолютно понятна. Если не понятна, войдите в неё отладчиком и отследите все действия, соотнесите их с расчётами на бумажке.

**Comment by kalininei** [comment]
Да, в цикле (9.5), как вы наверно уже поняли, умножение коэффициентов `v` на площадь грани пропущено. Надо его добавить.

**Comment by Poodelay** [comment]
![Снимок экрана от 2024-04-19 22-44-53](https://github.com/kalininei/CFDCourse25/assets/96184011/dc4accd6-2bf9-4f15-a0d6-4fb353cad0be)
![Снимок экрана от 2024-04-19 22-45-12](https://github.com/kalininei/CFDCourse25/assets/96184011/2c2e657a-b675-4b6e-b0ea-546bd7f1bfdf)
Извините, что поздно, но я что-то совсем не понимаю. На первом скриншоте код, который вы добавили в рекомендации по программированию плюс прописан вывод его в консоль, на втором скриншоте то что он выводит. Я использую для вывода структурированную двумерную сетку размера 10 на 10, и если правильно понимаю, то я аппроксимирую производную на грани между первой и второй ячейками значениями в двух разных углах всей расчетной области. Также в какой-то момент i и j принимают вообще одинаковые индексы

**Comment by kalininei** [comment]
Не понимаю, как у вас это получилось. Вот я ввёл ваши данные
![image](https://github.com/kalininei/CFDCourse25/assets/5234858/f6d690f3-47e6-4c08-ac42-29b14eb160ba)
Ответ такой:
![image](https://github.com/kalininei/CFDCourse25/assets/5234858/87c18a81-f1c1-4a38-a656-7048652c2ed8)

**Comment by kalininei** [comment]
А, теперь понимаю)
![image](https://github.com/kalininei/CFDCourse25/assets/5234858/bd1624c9-3e7f-44e5-9b92-6392fd24f941)

**Comment by Poodelay** [comment]
Да, тоже увидел, спасибо
